lib/composite_cache_store.rb

```diff
@@ -4,146 +4,142 @@
+    provisional_layers.each { |layer| layer.write(name, value, options) }
+    value
+  end
+
+  def decrement(name, amount = 1, options = nil)
+    value = layers.last.decrement(name, amount, options)
+    provisional_layers.each { |layer| layer.write(name, value, options) }
+    value
   end
 
   def delete(...)
-    layers.each { |store| store.delete(...) }
+    layers.each { |layer| layer.delete(...) }
   end
 
   def delete_matched(...)
-    layers.each { |store| store.delete_matched(...) }
+    layers.each { |layer| layer.delete_matched(...) }
   end
 
   def delete_multi(...)
-    layers.each { |store| store.delete_multi(...) }
+    layers.map { |layer| layer.delete_multi(...) }.last
   end
 
   def exist?(...)
-    layers.each do |store|
-      return true if store.exist?(...)
+    layers.each do |layer|
+      return true if layer.exist?(...)
     end
     false
   end
 
-  def fetch(*args, &block)
-    f = ->(store) do
-      return store.fetch(*args, &block) if store == layers.last
-      store.fetch(*args) { f.call(layers[layers.index(store) + 1]) }
+  def fetch(name, options = nil, &block)
+    value = read(name, options) unless options&.dig(:force) || options&.dig(:race_condition_ttl)
+    return value if value
+
+    layers.each do |layer|
+      if value
+        layer.write name, value, options
+      else
+        value = layer.fetch(name, options, &block)
+      end
     end
-    f.call(layers.first)
+
+    value
   end
 
   def fetch_multi(*args, &block)
-    fm = ->(store) do
-      return store.fetch_multi(*args, &block) if store == layers.last
-      store.fetch_multi(*args) { fm.call(layers[layers.index(store) + 1]) }
-    end
-    fm.call(layers.first)
-  end
+    names = args.dup
+    options = names.extract_options!
 
-  def increment(...)
-    layers.each { |store| store.increment(...) }
+    value = read_multi(*names) unless options[:force] || options[:race_condition_ttl]
+    value ||= {}
+    return value if value.compact.size == names.size
+
+    value = {}
+    layers.each do |layer|
+      if value.compact.size == names.size
+        layer.write_multi value, options
+      else
+        value = layer.fetch_multi(*args, &block)
+      end
+    end
+    value
   end
 
   def mute
-    m = ->(store) do
-      return store.mute { yield } if store == layers.last
-      store.mute { m.call(layers[layers.index(store) + 1]) }
-    end
-    m.call(layers.first)
+    layers.each { |layer| layer.mute { yield } }
   end
 
-  def read(*args)
-    r = ->(store) do
-      return store.read(*args) if store == layers.last
-      store.fetch(*args) { r.call(layers[layers.index(store) + 1]) }
+  def read(name, options = nil)
+    value = nil
+    cache_miss_layers = []
+    layers.find do |layer|
+      value = layer.read(name, options)
+      cache_miss_layers << layer unless value
+      value
     end
-    r.call(layers.first)
+    cache_miss_layers.each { |layer| layer.write(name, value, options) } if value
```

Comment:
I understand the intent here, but is this the proper semantics? Do we really want to exert side effects when reading from the cache (since this isn't `fetch`)

Reply:
Thanks for the push back. I'm certainly open to changing the behavior to avoid unexpected side effects, but given the purpose of lib I think this may be justified? 🤷🏻‍♂️ _Would also love to hear from others on this topic._

Reply:
This will be trivial to change to zero-side-effect reads if that's what we decide to do.